### PR TITLE
implement valhalla's trace_attributes

### PR DESCRIPTION
**Pull request by nilsnolde**
was a bit painful but should work now..

### CODE REVIEW

File: routingpy/routers/valhalla.py [hunk 1]
```diff
@@ -207,6 +210,7 @@ def directions(
             avoid_polygons,
             date_time,
             id,
+            **kwargs
```
Comment (nilsnolde) [inline routingpy/routers/valhalla.py:213]: kwargs didn't really exist before here either

File: routingpy/valhalla_attributes.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+from routingpy.utils import decode_polyline6
+
+
+class MatchDiscontinuity(str, Enum):
+    BEGIN = "begin"
+    END = "end"
+    NONE = ""
+
+
+class MatchType(str, Enum):
```
Comment (nilsnolde) [inline routingpy/valhalla_attributes.py:32]: these are fixed enums for response entries

File: routingpy/valhalla_attributes.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+    FERRY = "ferry"
+    RAIL = "rail"
+    BUS = "bus"
+    RAIL_CONNECTION = "rail_connection"
+    BUS_CONNECTION = "bus_connnection"
+    TRANSIT_CONNECTION = "transit_connection"
+    NONE = ""
+
+
+class MatchedEdge:
```
Comment (nilsnolde) [inline routingpy/valhalla_attributes.py:111]: this is a single edge object with most interesting properties set. could be extended in future, but good enough for NA

File: routingpy/valhalla_attributes.py [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+        """
+        The length of this edge in meters.
+        """
+        return self._length
+
+    def __repr__(self):  # pragma: no cover
+        return "Edge({})".format(", ".join([f"{k[1:]}: {v}" for k, v in vars(self).items() if v]))
+
+
+class MatchedPoint:
```
Comment (nilsnolde) [inline routingpy/valhalla_attributes.py:304]: same for matched points

File: routingpy/valhalla_attributes.py [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+
+    @property
+    def discontinuity(self) -> Optional[MatchDiscontinuity]:
+        """
+        Returns the :class:`MatchDiscontinuity` status.
+        """
+        return self._discontinuity
+
+
+class MatchedResults:
```
Comment (nilsnolde) [inline routingpy/valhalla_attributes.py:365]: this is what the user will get. he can then go through the `matched_edges` and `matched_points` lists

File: routingpy/valhalla_attributes.py [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+        if not response:
+            return
+
+        geometry = decode_polyline6(response["shape"])
+        # fill the edges
+        for edge in response["edges"]:
+            coords: List[List[float]] = geometry[
+                response.get("begin_shape_index")
+                or 0 : (response.get("end_shape_index") or (len(geometry) - 1)) + 1
+            ]
```
Comment (nilsnolde) [inline routingpy/valhalla_attributes.py:385]: begin/end_shape_index can be empty for first/last edge :roll_eyes:

File: CHANGELOG.md [hunk 7]
```diff
@@ -8,6 +8,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## **Unreleased**
 
+## [v1.1.0](https://pypi.org/project/routingpy/1.1.0/)
+
+### Fixed
+- Added Valhalla's trace_attributes endpoint
```
Comment (chrstnbwnkl) [inline CHANGELOG.md:14]: Shouldn't this be listed under `added`?

File: routingpy/routers/valhalla.py [hunk 8]
```diff
@@ -802,6 +824,102 @@ def parse_expansion_json(response, locations, expansion_properties, interval_typ
...
+        profile: str = "bicycle",
+        shape_match: str = "walk_or_snap",
+        encoded_polyline: Optional[str] = None,
+        filters: Optional[List[str]] = None,
+        filters_action: Optional[str] = None,
+        options: Optional[dict] = None,
+        dry_run: Optional[bool] = None,
+        **kwargs
+    ) -> MatchedResults:
+        """Gets the expansion tree for a range of time or distance values around a given coordinate.
```
Comment (chrstnbwnkl) [inline routingpy/routers/valhalla.py:839]: copied expansion docstring

**Review by chrstnbwnkl — CHANGES_REQUESTED** [review]
Almost good to go, just one faulty condition and a docstring that needs a small update

File: routingpy/valhalla_attributes.py [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+    def __init__(self, point: dict):
+        self._geometry: List[float] = [point["lon"], point["lat"]]
+        self._match_type = MatchType(point.get("type", "")) or None
+        self._dist_along_edge: Optional[float] = point.get("distance_along_edge")
+        self._dist_from_input: Optional[int] = point.get("distance_from_trace_point")
+        self._edge_index: Optional[int] = point.get("distance_along_edge")
+        self._discontinuity: Optional[MatchDiscontinuity] = None
+        if point.get("begin_route_discontinuity"):
+            self._discontinuity = MatchDiscontinuity("begin")
+        elif point.get("begin_route_discontinuity"):
```
[the remark this answers is not in the document]
Comment (nilsnolde) [reply routingpy/valhalla_attributes.py:318]: good catch!

**Pull request merged**